x/protorev/keeper/epoch_hook_test.go

```diff
@@ -0,0 +1,74 @@
+package keeper_test
+
+import (
+	"fmt"
+
+	"github.com/osmosis-labs/osmosis/v13/x/protorev/types"
+)
+
+func (suite *KeeperTestSuite) TestEpochHook() {
+	// All of the pools initialized in the setup function are available in keeper_test.go
+	// akash <-> types.OsmosisDenomination
+	// juno <-> types.OsmosisDenomination
+	// ethereum <-> types.OsmosisDenomination
+	// bitcoin <-> types.OsmosisDenomination
+	// canto <-> types.OsmosisDenomination
+	// and so on...
+	expectedToSee := make(map[string]Pool)
+	for _, pool := range suite.pools {
+
+		// Module currently limited to two asset pools
+		// Instantiate asset and amounts for the pool
+		if len(pool.PoolAssets) == 2 {
+			pool.Asset1 = pool.PoolAssets[0].Token.Denom
+			pool.Amount1 = pool.PoolAssets[0].Token.Amount
+			pool.Asset2 = pool.PoolAssets[1].Token.Denom
+			pool.Amount2 = pool.PoolAssets[1].Token.Amount
+		}
+
+		if pool.Asset1 == types.OsmosisDenomination || pool.Asset2 == types.OsmosisDenomination || pool.Asset1 == types.AtomDenomination || pool.Asset2 == types.AtomDenomination {
+			// create a key that is a combination of asset1 and asset2 in alphabetical order
+			key := fmt.Sprintf("%s-%s", pool.Asset1, pool.Asset2)
+			if pool.Asset1 > pool.Asset2 {
+				key = fmt.Sprintf("%s-%s", pool.Asset2, pool.Asset1)
+			}
+
+			if storedPool, ok := expectedToSee[key]; !ok {
+				expectedToSee[key] = pool
+			} else {
+				liquidity := pool.Amount1.Mul(pool.Amount2)
```

Comment:
Asking to learn: Why is liquidity Multiplied between two pools? Shouldn't we be adding?

Reply:
The module assumes the constant product market maker model (x * y = k) where x and y are the pool reserves and k is can be seen as total liquidity in the pool. You can read more about it [here](https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf)

Reply:
aah yes the cfmm equation!